### PR TITLE
Expose OperationPurpose via CustomQueryParameter to s3 logs

### PR DESCRIPTION
**Pull request by ywangd**
This PR adds the OperationPurpose as a custom query parameter for each S3 request so that they are available in s3 access logs.

Resolves: ES-7750

### CODE REVIEW

File: modules/repository-s3/src/internalClusterTest/java/org/elasticsearch/repositories/s3/S3BlobStoreRepositoryTests.java [hunk 1]
```diff
@@ -423,6 +424,21 @@ public void testEnforcedCooldownPeriod() throws IOException {
...
     }
 
+    private Map<S3BlobStore.StatsKey, Long> getServerMetrics() {
+        for (HttpHandler h : handlers.values()) {
+            while (h instanceof DelegatingHttpHandler) {
+                if (h instanceof S3StatsCollectorHttpHandler s3StatsCollectorHttpHandler) {
+                    return s3StatsCollectorHttpHandler.getMetricsCount()
+                        .entrySet()
+                        .stream()
+                        .collect(Collectors.toUnmodifiableMap(Map.Entry::getKey, entry -> entry.getValue().get()));
```
Comment (idegtiarenko) [inline modules/repository-s3/src/internalClusterTest/java/org/elasticsearch/repositories/s3/S3BlobStoreRepositoryTests.java:434]: NIT: could be simplified with `Maps.transformValues`

**Review by idegtiarenko — APPROVED** [review]

**Comment by ywangd** [comment]
@elasticmachine update branch

**Comment by ywangd** [comment]
@elasticmachine run elasticsearch-ci/part-1

**Comment by ywangd** [comment]
@elasticmachine update branch

**Pull request merged**